core/src/proof.rs

```diff
@@ -0,0 +1,135 @@
+//! Proving and verifying inclusion, non-inclusion, and updates to the trie.
+
+use crate::trie::{InternalData, KeyPath, LeafData, Node, NodeHasher, NodeHasherExt, TERMINATOR};
+
+use alloc::vec::Vec;
+use bitvec::prelude::*;
+
+/// Generic cursor over binary trie storage.
+pub trait Cursor {
+    /// The current position of the cursor, expressed as a bit-path and length. Bits after the
+    /// length are irrelevant.
+    fn position(&self) -> (KeyPath, u8);
+    /// The current node.
+    fn node(&self) -> Node;
+
+    /// Peek at the children of this node, if it is an internal node.
+    fn peek_children(&self) -> Option<(&Node, &Node)>;
```

Comment:
`&Node` forces the function to borrow the Node from `Self`. Would it be always possible?

Reply:
I think this is reasonable to expect for anything operating out of memory, and should be fine but I would be open to change it later on if needed.

Reply:
Right now, I am not 100% sure that this will be sufficient, but it also seems low-risk to merge.